aten/src/ATen/TensorNames.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include <ATen/core/EnableNamedTensor.h>
+#include <ATen/WrapDimUtils.h>
+
+namespace at { namespace namedinference {
+
+#ifdef BUILD_NAMEDTENSOR
+
+// TensorName and TensorNames are wrappers around Dimname and DimnameList
+// that contain helper functions to make writing name inference rules easier.
+//
+// A TensorName represents a Dimname associated with some DimnameList (from a Tensor).
+// This encapsulates all the information that is needed to check if names *match*
+// and to *unify* names.
+//
+// Definition: Two names in two tensors *match* if they are equal, or if at
+// least one of them is a wildcard that can be *refined* to the other name.
+//
+// Here is an example of checking if two names match.
+// tensor: Tensor[A, None]
+// other: Tensor[A]
+//
+// Let's say we wish to check if tensor.names[-1] matches other.names[-1].
+// None (in tensor) cannot match A (in other) because if the None were refined
+// to A, `tensor` would have duplicate names [A, A]. Therefore we need to check
+// tensor.names [A, None] for the existence of A.
```

Comment:
I like having this definition/explanation. I know it might seem obvious from the name but it might also be helpful to define 'unify' as well.

Reply:
Done